### PR TITLE
[MIR] Handle call return values that need to be casted properly.

### LINKED ISSUE #33873: MIR trans cannot handle casting for call return values.
In the following example, `bar` optimizes to `ret i8 undef` because the return value is cast into an `i32` in `foo` and has to be converted back at the call site. The old trans code does this with a gnarly `memcpy` which should handle all of the possible situations, if the comment on it is of any indication.
I believe my use of `ArgType::store` for storing the return value in its destination was unwarranted.
Since then, @Aatch has refactored that logic into its own function, which should do the `memcpy`.
cc @nagisa @dotdash

```rust
#![crate_type = "lib"]
#![feature(rustc_attrs)]

pub extern "C" fn foo() -> (u8, u8, u8) {
    (1, 2, 3)
}

#[rustc_mir]
pub fn bar() -> u8 {
    foo().2
}
```

### PR DESCRIPTION
**Pull request by luqmana**
Fixes #33873

r? @eddyb

### CODE REVIEW

File: src/librustc_trans/mir/block.rs [hunk 1]
```diff
@@ -685,7 +687,37 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
+            Store(dst) => {
+                if let Some(llcast_ty) = ret_ty.cast {
+                    let ccx = bcx.ccx();
+                    // The actual return type is a struct, but the ABI
+                    // adaption code has cast it into some scalar type.
+                    // We can't use `ret_ty.store(..)` since that simply
+                    // does a `PointerCast` and a `Store`. Unfortunately,
+                    // that may run afoul of strict aliasing rules, leading
+                    // to wacky results after optimization since LLVM
+                    // may just replace it with an `undef`.
```
Comment (eefriedman) [inline src/librustc_trans/mir/block.rs]: The problem has nothing to do with aliasing rules... the problem is the buffer overflow.

**Comment by eddyb** [comment]
LGTM but to be sure r? @dotdash

File: src/librustc_trans/mir/block.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_trans/mir/block.rs]: I'd prefer copying the comment from `callee`.

**Comment by dotdash** [comment]
Seems ok except for the misleading comment. But you might actually want to extract that into some function (not sure if `store` is the right candidate), because we still need a similar for passing arguments to C ABI functions (either that or tighter type calculations for C ABI functions).

For example:
```rust
#![crate_type="lib"]
struct S {
    x: i8,
    y: i8,
}

extern {
    fn foo(s: S);
}

pub fn bla() {
    unsafe {
        foo(S { x: 4, y: 8 });
    }
}
```

Results in:
```llvm
define void @_ZN3ffi3bla17h4acc14686285caf0E() unnamed_addr #0 {
entry-block:
  %temp0 = alloca {}
  %temp1 = alloca %S
  %arg = alloca %S
  br label %start

start:                                            ; preds = %entry-block
  %0 = getelementptr inbounds %S, %S* %temp1, i32 0, i32 0
  store i8 4, i8* %0
  %1 = getelementptr inbounds %S, %S* %temp1, i32 0, i32 1
  store i8 8, i8* %1
  %2 = load %S, %S* %temp1
  store %S %2, %S* %arg
  %3 = bitcast %S* %arg to { i64 }*
  %4 = load { i64 }, { i64 }* %3, align 1
  call void @foo({ i64 } %4)
  br label %bb2

end:                                              ; preds = %bb3
  ret void

bb2:                                              ; preds = %start
  br label %bb3

bb3:                                              ; preds = %bb2
  br label %end
}
```

Notice how we try to load 8 bytes out of a 2 byte allocation.

Also, we perform FCA loads/store :-( (that's unrelated to this PR though).

**Comment by eddyb** [comment]
```

failures:

---- [codegen] codegen/stores.rs stdout ----
	
error: verification with 'FileCheck' failed
status: exit code: 1
command: /usr/lib/llvm-3.7/bin/FileCheck -input-file=x86_64-unknown-linux-gnu/test/codegen/stores.ll /build/src/test/codegen/stores.rs
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
/build/src/test/codegen/stores.rs:29:11: error: expected string not found in input
// CHECK: store i32 %{{.*}}, i32* %{{.*}}, align 1
          ^
x86_64-unknown-linux-gnu/test/codegen/stores.ll:8:35: note: scanning from here
define void @small_array_alignment([4 x i8]* dereferenceable(4), i32) unnamed_addr #0 {
                                  ^
x86_64-unknown-linux-gnu/test/codegen/stores.ll:15:2: note: possible intended match here
 store i32 %1, i32* %fn_ret_cast
 ^
/build/src/test/codegen/stores.rs:40:11: error: expected string not found in input
// CHECK: store i32 %{{.*}}, i32* %{{.*}}, align 1
          ^
x86_64-unknown-linux-gnu/test/codegen/stores.ll:44:36: note: scanning from here
define void @small_struct_alignment(%Bytes* dereferenceable(4), i32) unnamed_addr #0 {
                                   ^
x86_64-unknown-linux-gnu/test/codegen/stores.ll:51:2: note: possible intended match here
 store i32 %1, i32* %fn_ret_cast
 ^
```

r=me with this test failure fixed

**Comment by eddyb** [comment]
FWIW I've included a form of this in #33905, because it needs some extra logic to also work for the generalized `Pair` operands.
If that doesn't get merged first, the additions are in a separate commit, so it should work out either way.

**Pull request closed without merging**